File: agent/collectors/asana.py
"""
Asana collector.

Fetches incomplete tasks assigned to Nat that are due by end of weekend
(or have no due date — those are shown as "no deadline" reminders).

Uses the Asana REST API directly via httpx (avoids SDK version churn).
Auth: Personal Access Token stored as ASANA_PAT secret.
"""

from datetime import date
from zoneinfo import ZoneInfo

EASTERN = ZoneInfo("America/New_York")

import httpx

from agent.config import settings
from agent.models import AsanaTask

ASANA_BASE = "https://app.asana.com/api/1.0"


def _headers() -> dict[str, str]:
    return {"Authorization": f"Bearer {settings.asana_pat}"}


def _fetch_tasks_for_assignee(
    client: httpx.Client,
    assignee: str,
    workspace_gid: str,
    week_end: date,
) -> list[AsanaTask]:
    """Fetch incomplete tasks for a given assignee GID (or 'me')."""
    params: dict[str, str] = {
        "assignee": assignee,
        "workspace": workspace_gid,
        "completed_since": "now",
        "opt_fields": "gid,name,due_on,notes,memberships.project.name,permalink_url,assignee.gid,assignee.name",
        "limit": "100",
    }
    url = f"{ASANA_BASE}/tasks"
    tasks: list[AsanaTask] = []

    while True:
        resp = client.get(url, headers=_headers(), params=params)
        try:
            resp.raise_for_status()
        except httpx.HTTPStatusError as e:
            print(f"[asana] HTTP error {e.response.status_code}: {e.response.text}")
            break

        data = resp.json()
        for item in data.get("data", []):
            due_on = item.get("due_on")
            due_date = date.fromisoformat(due_on) if due_on else None
            memberships = item.get("memberships") or []
            project_name: str | None = None
            if memberships:
                project = memberships[0].get("project") or {}
                project_name = project.get("name")
            raw_notes = item.get("notes") or ""
            assignee = item.get("assignee") or {}
            tasks.append(
                AsanaTask(
                    gid=item["gid"],
                    name=item["name"],
                    due_on=due_date,
                    project=project_name,
                    url=item.get(
                        "permalink_url",
                        f"https://app.asana.com/0/0/{item['gid']}",
                    ),
                    notes=raw_notes[:200] if raw_notes else None,
                    assignee_name=assignee.get("name"),
                )
            )

        next_page = data.get("next_page")
        if next_page and next_page.get("offset"):
            params["offset"] = next_page["offset"]
        else:
            break

    return [t for t in tasks if t.due_on is not None and t.due_on <= week_end]


def fetch_week_tasks(week_end: date) -> list[AsanaTask]:
    """
    Fetch incomplete tasks assigned to me that are past due, due today, or due
    within the coming week (up to week_end). Tasks with no due date are excluded.

    'completed_since=now' returns only incomplete tasks (Asana API convention).
    """
    if settings.asana_project_gid:
        # Project-scoped path (no assignee filter)
        params: dict[str, str] = {
            "completed_since": "now",
            "opt_fields": "gid,name,due_on,notes,memberships.project.name,permalink_url,assignee.gid,assignee.name",
            "limit": "100",
        }
        url = f"{ASANA_BASE}/projects/{settings.asana_project_gid}/tasks"
        tasks: list[AsanaTask] = []
        with httpx.Client(timeout=20.0) as client:
            while True:
                resp = client.get(url, headers=_headers(), params=params)
                try:
                    resp.raise_for_status()
                except httpx.HTTPStatusError as e:
                    print(f"[asana] HTTP error {e.response.status_code}: {e.response.text}")
                    break
                data = resp.json()
                for item in data.get("data", []):
                    due_on = item.get("due_on")
                    due_date = date.fromisoformat(due_on) if due_on else None
                    memberships = item.get("memberships") or []
                    project_name: str | None = None
                    if memberships:
                        project = memberships[0].get("project") or {}
                        project_name = project.get("name")
                    raw_notes = item.get("notes") or ""
                    tasks.append(
                        AsanaTask(
                            gid=item["gid"],
                            name=item["name"],
                            due_on=due_date,
                            project=project_name,
                            url=item.get(
                                "permalink_url",
                                f"https://app.asana.com/0/0/{item['gid']}",
                            ),
                            notes=raw_notes[:200] if raw_notes else None,
                        )
                    )
                next_page = data.get("next_page")
                if next_page and next_page.get("offset"):
                    params["offset"] = next_page["offset"]
                else:
                    break
        tasks = [t for t in tasks if t.due_on is not None and t.due_on <= week_end]
    else:
        with httpx.Client(timeout=20.0) as client:
            tasks = _fetch_tasks_for_assignee(client, "me", settings.asana_workspace_gid, week_end)

    tasks.sort(key=lambda t: t.due_on)
    return tasks


def fetch_workspace_tasks(week_end: date) -> list[AsanaTask]:
    """
    Fetch incomplete tasks for ALL users in the workspace.
    Used by the TRMNL display to show both Nat's and Caitie's tasks.
    Deduplicates by task GID.
    """
    workspace_gid = settings.asana_workspace_gid

    with httpx.Client(timeout=20.0) as client:
        resp = client.get(
            f"{ASANA_BASE}/workspaces/{workspace_gid}/users",
            headers=_headers(),
            params={"opt_fields": "gid,name"},
        )
        try:
            resp.raise_for_status()
        except httpx.HTTPStatusError as e:
            print(f"[asana] Could not list workspace users ({e.response.status_code}), falling back to 'me'")
            tasks = _fetch_tasks_for_assignee(client, "me", workspace_gid, week_end)
            tasks.sort(key=lambda t: t.due_on)
            return tasks

        users = resp.json().get("data", [])
        if not users:
            tasks = _fetch_tasks_for_assignee(client, "me", workspace_gid, week_end)
            tasks.sort(key=lambda t: t.due_on)
            return tasks

        seen: set[str] = set()
        all_tasks: list[AsanaTask] = []
        for user in users:
            print(f"[asana] Fetching tasks for user {user.get('name', user['gid'])}")
            for task in _fetch_tasks_for_assignee(client, user["gid"], workspace_gid, week_end):
                if task.gid not in seen:
                    seen.add(task.gid)
                    all_tasks.append(task)

    all_tasks.sort(key=lambda t: t.due_on)
    return all_tasks

File: shared/src/asana.ts
import { format, addDays, subDays } from 'date-fns'
import type { AsanaTask, AsanaUser } from './types.ts'

const BASE = 'https://app.asana.com/api/1.0'

export type AsanaConfig = {
  pat: string
  workspaceGid: string
}

export type AsanaClient = {
  fetchTasks: () => Promise<AsanaTask[]>
  fetchAllOpenTasks: () => Promise<AsanaTask[]>
  fetchWorkspaceUsers: () => Promise<AsanaUser[]>
  fetchMe: () => Promise<AsanaUser>
  createTask: (fields: { name: string; due_on?: string; assignee?: string; notes?: string }) => Promise<AsanaTask>
  updateTask: (
    gid: string,
    fields: Partial<{ name: string; due_on: string | null; assignee: string | null; notes: string | null; completed: boolean }>,
  ) => Promise<AsanaTask>
  deleteTask: (gid: string) => Promise<void>
}

const TASK_OPT_FIELDS = 'gid,name,due_on,completed,completed_at,assignee.gid,assignee.name,memberships.project.name,notes'

function withTaskOptFields(path: string): string {
  return path.includes('?') ? `${path}&opt_fields=${TASK_OPT_FIELDS}` : `${path}?opt_fields=${TASK_OPT_FIELDS}`
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function parseTask(raw: any): AsanaTask {
  const projects: string[] = []
  if (Array.isArray(raw.memberships)) {
    for (const m of raw.memberships) {
      if (m.project?.name) projects.push(m.project.name as string)
    }
  }
  return {
    gid: raw.gid as string,
    name: raw.name as string,
    due_on: (raw.due_on as string | null) ?? null,
    completed: raw.completed as boolean,
    completed_at: (raw.completed_at as string | null) ?? null,
    assignee: raw.assignee
      ? { gid: raw.assignee.gid as string, name: raw.assignee.name as string }
      : null,
    notes: (raw.notes as string | null) || null,
    projects,
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function parseUsers(json: any): AsanaUser[] {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (json.data as any[]).map(u => ({
    gid: u.gid as string,
    name: u.name as string,
    email: (u.email as string | undefined) ?? '',
  }))
}

export function createAsanaClient(config: AsanaConfig): AsanaClient {
  const { pat, workspaceGid } = config

  // Cached after first successful resolveWorkspace() call — used by createTask
  let _resolvedWorkspaceGid: string | null = null
  // Full resolution cache — resolveWorkspace() is called by fetchTasks,
  // fetchAllOpenTasks and fetchWorkspaceUsers, so memoize to run it once.
  let _resolved: { workspaceGid: string; users: AsanaUser[] } | null = null

  function headers() {
    return {
      Authorization: `Bearer ${pat}`,
      'Content-Type': 'application/json',
    }
  }

  async function asanaGet(path: string): Promise<unknown> {
    const res = await fetch(`${BASE}${path}`, { headers: headers() })
    if (!res.ok) throw new Error(`Asana API error ${res.status}: ${await res.text()}`)
    return res.json()
  }

  async function asanaPost(path: string, body: unknown): Promise<unknown> {
    const res = await fetch(`${BASE}${path}`, {
      method: 'POST',
      headers: headers(),
      body: JSON.stringify({ data: body }),
    })
    if (!res.ok) throw new Error(`Asana API error ${res.status}: ${await res.text()}`)
    return res.json()
  }

  async function asanaPut(path: string, body: unknown): Promise<unknown> {
    const res = await fetch(`${BASE}${path}`, {
      method: 'PUT',
      headers: headers(),
      body: JSON.stringify({ data: body }),
    })
    if (!res.ok) throw new Error(`Asana API error ${res.status}: ${await res.text()}`)
    return res.json()
  }

  async function asanaDelete(path: string): Promise<void> {
    const res = await fetch(`${BASE}${path}`, { method: 'DELETE', headers: headers() })
    if (!res.ok) throw new Error(`Asana API error ${res.status}: ${await res.text()}`)
  }

  /**
   * Lists the members of a single workspace.
   *
   * NOTE: `/workspaces/{gid}/users` does NOT paginate. For a large org (e.g. an
   * employer's enterprise workspace) Asana rejects the whole request with
   * 400 "The result is too large." That's the signal that `gid` is the wrong
   * workspace — the configured one should be a small personal workspace, not a
   * 50k-user org. The caller treats any failure here as "skip this workspace."
   */
  async function listWorkspaceUsers(gid: string): Promise<AsanaUser[]> {
    return parseUsers(await asanaGet(`/workspaces/${gid}/users?opt_fields=gid,name,email`))
  }

  /**
   * Resolves the workspace whose users + tasks we fetch, memoized for the
   * client's lifetime.
   *
   * Happy path: the configured workspace listing succeeds, and we return without
   * ever calling `/workspaces` discovery or touching any other workspace the
   * account belongs to (which may be a huge org that 400s on user listing).
   * Only if the configured workspace is unusable do we discover + probe others,
   * then fall back to the PAT owner alone.
   *
   * Critical: the returned workspaceGid MUST be used for task fetching — user
   * GIDs are workspace-scoped, so mixing workspaces causes 404 on /tasks.
   */
  async function resolveWorkspace(): Promise<{ workspaceGid: string; users: AsanaUser[] }> {
    if (_resolved) return _resolved

    // 1. Configured workspace first — the expected happy path.
    try {
      const users = await listWorkspaceUsers(workspaceGid)
      _resolvedWorkspaceGid = workspaceGid
      _resolved = { workspaceGid, users }
      return _resolved
    } catch { /* unusable (wrong/too-large) — fall through to discovery */ }

    // 2. Discover other workspaces and probe each; skip any that fail.
    let discovered: string[] = []
    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const ws = await asanaGet('/workspaces?opt_fields=gid') as any
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      discovered = (ws.data as any[]).map(w => w.gid as string).filter(g => g !== workspaceGid)
    } catch { /* ignore */ }

    for (const gid of discovered) {
      try {
        const users = await listWorkspaceUsers(gid)
        _resolvedWorkspaceGid = gid
        _resolved = { workspaceGid: gid, users }
        return _resolved
      } catch { continue }
    }

    // 3. Final fallback: just the PAT owner; use configured workspace for tasks.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const me = await asanaGet('/users/me?opt_fields=gid,name,email') as any
    _resolvedWorkspaceGid = workspaceGid
    _resolved = {
      workspaceGid,
      users: [{
        gid: me.data.gid as string,
        name: me.data.name as string,
        email: (me.data.email as string | undefined) ?? '',
      }],
    }
    return _resolved
  }

  async function fetchTasksForUser(userGid: string, wsGid: string, completedSince: string): Promise<AsanaTask[]> {
    const all: AsanaTask[] = []
    let offset: string | null = null

    do {
      const params = new URLSearchParams({
        assignee: userGid,
        workspace: wsGid,
        completed_since: completedSince,
        opt_fields: 'gid,name,due_on,completed,completed_at,assignee.gid,assignee.name,memberships.project.name,notes',
        limit: '100',
      })
      if (offset) params.set('offset', offset)

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const json = await asanaGet(`/tasks?${params.toString()}`) as any
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      for (const raw of (json.data as any[])) {
        all.push(parseTask(raw))
      }
      offset = json.next_page?.offset ?? null
    } while (offset)

    return all
  }

  /**
   * Fetches tasks for all allowed users.
   * Returns:
   *   - Incomplete tasks with due_on ≤ today+7 (past due, today, this week)
   *   - Tasks completed within the last 7 days (for the "recently completed" section)
   */
  async function fetchTasks(): Promise<AsanaTask[]> {
    const { workspaceGid: resolvedGid, users: allUsers } = await resolveWorkspace()

    const userGids = allUsers.map(u => u.gid)
    if (userGids.length === 0) userGids.push('me')

    const cutoff = format(addDays(new Date(), 7), 'yyyy-MM-dd')
    const sevenDaysAgo = subDays(new Date(), 7).toISOString()

    const seen = new Set<string>()
    const all: AsanaTask[] = []

    for (const gid of userGids) {
      const tasks = await fetchTasksForUser(gid, resolvedGid, sevenDaysAgo)
      for (const t of tasks) {
        if (!seen.has(t.gid)) {
          seen.add(t.gid)
          all.push(t)
        }
      }
    }

    return all.filter(t =>
      t.completed ||
      (!t.completed && t.due_on !== null && t.due_on <= cutoff)
    )
  }

  /**
   * Fetches every incomplete task assigned to any allowed user in the workspace,
   * regardless of due date. Used by dashboard search to surface tasks outside the
   * normal "due ≤ today+7" rendering window.
   *
   * Asana's `/tasks` endpoint requires `completed_since` whenever you query by
   * assignee, so we pass epoch 0 and filter `completed === false` on the way out.
   */
  async function fetchAllOpenTasks(): Promise<AsanaTask[]> {
    const { workspaceGid: resolvedGid, users: allUsers } = await resolveWorkspace()

    const userGids = allUsers.map(u => u.gid)
    if (userGids.length === 0) userGids.push('me')

    const epoch = new Date(0).toISOString()

    const seen = new Set<string>()
    const all: AsanaTask[] = []

    for (const gid of userGids) {
      const tasks = await fetchTasksForUser(gid, resolvedGid, epoch)
      for (const t of tasks) {
        if (!seen.has(t.gid) && !t.completed) {
          seen.add(t.gid)
          all.push(t)
        }
      }
    }

    return all
  }

  async function fetchWorkspaceUsers(): Promise<AsanaUser[]> {
    const { users } = await resolveWorkspace()
    return users
  }

  async function fetchMe(): Promise<AsanaUser> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const me = await asanaGet('/users/me?opt_fields=gid,name,email') as any
    return {
      gid: me.data.gid as string,
      name: me.data.name as string,
      email: (me.data.email as string | undefined) ?? '',
    }
  }

  async function createTask(fields: {
    name: string
    due_on?: string
    assignee?: string
    notes?: string
  }): Promise<AsanaTask> {
    const wsGid = _resolvedWorkspaceGid ?? workspaceGid
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const json = await asanaPost(withTaskOptFields('/tasks'), { ...fields, workspace: wsGid }) as any
    return parseTask(json.data)
  }

  async function updateTask(
    gid: string,
    fields: Partial<{ name: string; due_on: string | null; assignee: string | null; notes: string | null; completed: boolean }>,
  ): Promise<AsanaTask> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const json = await asanaPut(withTaskOptFields(`/tasks/${gid}`), fields) as any
    return parseTask(json.data)
  }

  async function deleteTask(gid: string): Promise<void> {
    await asanaDelete(`/tasks/${gid}`)
  }

  return {
    fetchTasks,
    fetchAllOpenTasks,
    fetchWorkspaceUsers,
    fetchMe,
    createTask,
    updateTask,
    deleteTask,
  }
}
